lib/fast_jsonapi/serialization_core.rb

```diff
@@ -65,46 +66,90 @@ def meta_hash(record, params = {})
         FastJsonapi.call_proc(meta_to_serialize, record, params)
       end
 
+      def generate_cache_key(record, namespace)
+        "#{namespace}-#{record.cache_key}"
+      end
+
       def record_hash(record, fieldset, includes_list, params = {})
         if cache_store_instance
-          cache_opts = record_cache_options(cache_store_options, fieldset, includes_list, params)
-          record_hash = cache_store_instance.fetch(record, **cache_opts) do
-            temp_hash = id_hash(id_from_record(record, params), record_type, true)
-            temp_hash[:attributes] = attributes_hash(record, fieldset, params) if attributes_to_serialize.present?
-            temp_hash[:relationships] = relationships_hash(record, cachable_relationships_to_serialize, fieldset, includes_list, params) if cachable_relationships_to_serialize.present?
-            temp_hash[:links] = links_hash(record, params) if data_links.present?
-            temp_hash
+          cache_opts = record_cache_options(record, cache_store_options, fieldset, includes_list, params)
+          name_space = cache_opts.delete(:namespace)
+          cache_key = generate_cache_key(record, name_space)
+          klass = self
+          param = {
+            cache_key: cache_key,
+            cache_opts: cache_opts,
+            record: record,
+            klass: klass,
+            params: params
+          }
+          ::BatchLoader.for(param).batch do |batch_params, loader|
+            cache_keys = batch_params.map { |h| h[:cache_key] }
+            # load the cached value from cache store
+            cache_hits = cache_store_instance.read_multi(cache_keys)
+
+            # for not cached record, group it by cache options
+            uncached_by_opts = batch_params
+              .reject { |h| cache_hits.keys.include? h[:cache_key] }
+              .group_by { |h| h[:cache_opts] }
+            # replace grouped batch param with Hash: { cache_key_1 => record_hash_1, ... }
```

Comment:
this comment will be more useful if we show the overall structure of uncached_by_opts instead of only the value:
```
{
  {...cache_option} => {
    "cache_key_1" => {...record_hash_1},
    "cache_key_2"=> {...record_hash_2},
    ...
  },
  .... # different cache options
}
```